### PR TITLE
Handle requests with very large payload

### LINKED ISSUE #1754: Requests with very large payload (>1Gb) need to be gracefully handled
**Describe the bug**
Currently, there seems to be no validation that gracefully handles the case with very large payload (>1Gb are possible). Regardless of payload size, the Broker attempts to load it in-memory, which in some cases makes it prone to out-of-memory issues.

 **Expected behavior**
There should be a cap on payload. The Broker shouldn't attempt to process requests with payload exceeding the cap, but raise a client error (i.e. a response code like 4xx). It probably makes sense to make this cap configurable.

 **To Reproduce**
Both of the following tests resulted in a repro, although this must also be reproducible under a lower RPS.
**Test 1:** Generate the load of ~100 RPS with 3.5Gb request payload.
**Test 2:** Generate the load of ~1000 RPS with 256Mb request payload.

 **Exit critera**
A unit test validating this behavior.

### PR DESCRIPTION
**Pull request by Ectelion**
Fixes #1754. In particular:
- Prevents potential out-of-memory issues caused by requests with very large  payload;
- Addresses a functional issue that occurs even for relatively small payload sizes that exceeds PubSub limit on message size, which is 10Mb (non-extendable). Requests with payloads above that are not passing the publishing stage;
    * You may find it on [PubSub quotas page](https://cloud.google.com/pubsub/quotas): after the main table, there is a section with non-extendable quotas;
 

### On memory consumption
Memory stress test was re-run several times. Here's is how a typical memory consumption pattern looks like:
image
In both cases, there is an initial spike varying around 600-800MiB, which then gradually smoothes out. There appears to be little to no impact on memory caused by the MaxBytesReader wrapper. This can also be seen from [MaxBytesReader's implementation](https://github.com/golang/go/blob/2cd2ff6f564dce5be0c4fb7f06338ff7af3fc9a9/src/net/http/request.go#L1121), which turns out to be lightweight and does not have any notable memory overheads.

### CODE REVIEW

File: pkg/broker/ingress/handler.go [hunk 1]
```diff
@@ -47,6 +47,9 @@ const (
 	// TODO(liu-cong) configurable timeout
 	decoupleSinkTimeout = 30 * time.Second
 
+	// Limit for request payload in bytes (100Mb)
+	maxRequestBodyBytes = 100000000
```
Comment (Ectelion) [inline pkg/broker/ingress/handler.go:51]: Let me know if we'd rather have this as an env variable.

Comment (liu-cong) [reply pkg/broker/ingress/handler.go:51]: Can you document how this number is picked? Is this the same as the pubsub message size limit?

**Comment by ian-mi** [comment]
I recommend also wrapping the body in an io.LimitedReader.

File: pkg/broker/ingress/handler.go (hunk 1 above)
Comment (Ectelion) [reply pkg/broker/ingress/handler.go:51]: Sure, I have updated the description.

**Comment by Ectelion** [comment]
> I recommend also wrapping the body in an io.LimitedReader.

Some more context on this: I think this is indeed a more reliable approach. I myself considered MaxBytesReader, which is similar. In the context of recently completed perf tuning work (#1600), I wondered if the influence on memory consumption can be assumed to be negligible or whether the memory analysis in #1600 would need to be re-evaluated. Given this and that the malicious use case implies a wider problem (unauthorized access to the Broker first place), I thought it may be fine to keep it with current approach. What do you think?

**Review by grantr — APPROVED** [review]
Thanks @Ectelion!

/lgtm

**Pull request merged**